AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -796,17 +798,142 @@ - (BOOL)shouldCalculateLayoutWithConstrainedSize:(ASSizeRange)constrainedSize pa
   return _calculatedDisplayNodeLayout->isValidForConstrainedSizeParentSize(constrainedSize, parentSize) == NO;
 }
 
-- (ASLayoutElementType)layoutElementType
+// TODO: coalayout: Deprecate API instead call -[ASDisplayNode setNeedsLayout]
+- (void)invalidateCalculatedLayout
 {
-  return ASLayoutElementTypeDisplayNode;
+  [self __setNeedsLayout];
 }
 
-- (BOOL)canLayoutAsynchronous
+/**
+ * Invalidate the current calculated layout so it can be recalculated in the next layout pass
+ */
+- (void)__setNeedsLayout
 {
-  return !self.isNodeLoaded;
+  ASDN::MutexLocker l(__instanceLock__);
+
+  _calculatedDisplayNodeLayout->invalidate();
+  _pendingDisplayNodeLayout->invalidate();
 }
 
-#pragma mark - Automatic Hierarchy
+/**
+ * Returns a BOOL indicating whether the layer has been marked as needing a layout update.
+ */
+- (BOOL)__needsLayout
+{
+  return _calculatedDisplayNodeLayout->isDirty();
+}
+
+/**
+ * The node's supernodes are traversed until a ancestor node is found that does not require layout. Then layout
```

Comment:
This seems like the right traversal behavior to me.  I'm nearly certain that UIKit / CA also do this - that is why they internally have a different method called layoutBelowIfNeeded, since that version doesn't consider upward-traversal.  (we shouldn't add layoutBelow now, hopefully we never need it :) )